chore(devex): Migrate Error Tracking Models to Products Structure

webjunkie opened the pull request:
Migrated error tracking models from `posthog/models/error_tracking/` to `products/error_tracking/backend/` as part of the modularization effort to better organize PostHog's codebase into product-focused directories.

### What Changed

#### 📁 **File Structure Migration**
**From:**
```
posthog/models/error_tracking/
├── error_tracking.py
├── hogvm_stl.py
├── sql.py
└── test/
    └── test_error_tracking.py
```

**To:**
```
products/error_tracking/backend/
├── models.py  (combined from all model files)
├── hogvm_stl.py
├── sql.py
└── test/
    └── test_error_tracking.py
```

#### 🔄 **Import Updates Across Codebase**
All imports were automatically updated:
- `from posthog.models.error_tracking import ErrorTrackingIssue` → `from products.error_tracking.backend.models import ErrorTrackingIssue`
- `from posthog.models.error_tracking.sql import ...` → `from products.error_tracking.backend.sql import ...`
- `from posthog.models.error_tracking.hogvm_stl import ...` → `from products.error_tracking.backend.hogvm_stl import ...`

#### 🔗 **Foreign Key References**
Converted direct class references to string-based references for better app separation:
- `ForeignKey(Team, ...)` → `ForeignKey("posthog.Team", ...)`
- `ForeignKey(User, ...)` → `ForeignKey("posthog.User", ...)`
- `ForeignKey(Role, ...)` → `ForeignKey("ee.Role", ...)` (Role is in the ee app)

#### 💾 **Database Compatibility**
- All model Meta classes retain their `db_table` settings
- No database migrations required - tables remain unchanged

#### ⚙️ **Django Configuration**
- Error tracking app remains registered in Django
- Models are now loaded from `products.error_tracking.backend` instead of `posthog.models.error_tracking`
- Admin classes properly moved and registered

### Models Moved
- `ErrorTrackingIssue`
- `ErrorTrackingIssueFingerprintV2`
- `ErrorTrackingStackFrame`
- `ErrorTrackingSymbolSet`
- `ErrorTrackingIssueAssignment`
- `ErrorTrackingAssignmentRule`
- `ErrorTrackingGroupingRule`
- `ErrorTrackingSuppressionRule`
- `ErrorTrackingRelease`
- `ErrorTrackingExternalReference`
- `ErrorTrackingIssueFingerprint`

### Impact
- ✅ **No breaking changes** - All existing functionality preserved
- ✅ **No database migrations needed** - Table names unchanged
- ✅ **Backward compatible** - All imports automatically updated
- ✅ **Better code organization** - Error tracking is now a self-contained product module
- ✅ **Cleaner dependency graph** - Reduced cross-module dependencies through string-based FK references

### Files Affected
- **Moved**: 4 files from `posthog/models/error_tracking/`
- **Updated imports**: ~20+ files across the codebase including:
  - API endpoints (`posthog/api/error_tracking.py`)
  - Management commands (`posthog/management/commands/sync_exceptions_to_clickhouse.py`)
  - Query runners (`posthog/hogql_queries/error_tracking_query_runner.py`)
  - DAGs (`dags/symbol_set_cleanup.py`)
  - Tests and other dependencies

This migration establishes error tracking as a proper product module while maintaining full backward compatibility and requiring no database changes.

webjunkie committed:
Run model migration on error tracking

github-actions[bot] committed:
Update query snapshots

orian reviewed (approved):
Ok for ClickHouse.

daibhin commented on posthog/migrations/0867_initial_migration.py:
```diff
@@ -0,0 +1,195 @@
+    operations = [
+        migrations.SeparateDatabaseAndState(
+            database_operations=[
+                migrations.RunPython(update_content_type, reverse_content_type),
```
I don't really understand what this does / why it's necessary but looks reasonable

daibhin reviewed (commented):
Left a couple of comments, mainly being overly cautious because it's a relatively risky change. Once you're happy that they're all addressed I can run the migrations locally as a final sanity check

webjunkie committed:
Merge master into error-tracking migration branch

webjunkie committed:
chore(models): migrate error_tracking models to products app

Move error_tracking models from posthog/models/ to products/error_tracking/backend/ to align with product-based architecture.

Generated migrations handle the model transfer using SeparateDatabaseAndState pattern. Updated all imports across the codebase using LibCST.

webjunkie committed:
Merge branch 'chore/models-migrations-error-tracking-fresh' into chore/models-migrations-error-tracking

webjunkie committed:
chore(migrations): remove stale migration files from previous attempt

webjunkie committed:
chore: remove unrelated CI comparison scripts

webjunkie committed:
chore(error_tracking): renumber migration to 0878 after master merge

Master introduced 0877_delete_named_query_from_state while we had
0877_migrate_error_tracking_models. Renamed migration to 0878 and
updated dependencies accordingly.

webjunkie committed:
Merge remote-tracking branch 'origin/master' into chore/models-migrations-error-tracking

webjunkie committed:
fix: restore send_discussions_mentioned that was accidentally removed by migration script

webjunkie committed:
Merge commit '2022d4a39b' into chore/models-migrations-error-tracking

webjunkie committed:
chore(models): migrate error_tracking models to products app

webjunkie committed:
Merge fresh error_tracking migration into PR branch

webjunkie committed:
Merge remote-tracking branch 'origin/master' into chore/models-migrations-error-tracking

webjunkie committed:
fix(error_tracking): update api/__init__.py import after migration

webjunkie committed:
fix(error_tracking): rename API test file to avoid pytest conflict

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

webjunkie committed:
fix(error_tracking): remove unused imports after migration

webjunkie committed:
Merge branch 'master' into chore/models-migrations-error-tracking

webjunkie committed:
fix(tests): use CASCADE for TransactionTestCase database flush

Override _fixture_teardown() in NonAtomicBaseTest to pass allow_cascade=True
when flushing test databases. Required for models moved between Django apps,
as PostgreSQL needs CASCADE to truncate tables with FK constraints across
app boundaries.

Without this, TransactionTestCase cleanup fails when models reference each
other across different Django apps (e.g., error_tracking models referencing
posthog models).

webjunkie committed:
chore(models): migrate error_tracking models to products app

Move error_tracking models from posthog.models.error_tracking to products.error_tracking.backend.
Uses SeparateDatabaseAndState migration to preserve database tables while moving Django app ownership.

Changes:
- Move models to products/error_tracking/backend/models.py
- Move API files to products/error_tracking/backend/api/
- Move HogQL query runners to products/error_tracking/backend/hogql_queries/
- Update all imports across codebase
- Fix ForeignKey references (User, Role) with proper app labels
- Generate Django migration 0878_migrate_error_tracking_models

webjunkie committed:
fix(error_tracking): clean up migration artifacts

Update stale imports and exports after error_tracking migration:
- Fix imports in decide.py, remote_config.py, and query runner
- Remove error_tracking exports from posthog/models/__init__.py
- Update test snapshot paths to products/error_tracking/backend
- Add error_tracking module to tach.toml
- Add CASCADE fix to TransactionTestCase for cross-app FK constraints

webjunkie committed:
fix(error_tracking): use correct migration from PR branch

Use migration 0878 with ContentType updates from PR branch.
This properly updates Django contenttypes when moving models between apps.

webjunkie committed:
chore(error_tracking): move hogql query tests to products app

Move test files from posthog/hogql_queries/test to products/error_tracking/backend/hogql_queries/test.
Update imports to reference new paths.

webjunkie committed:
chore(error_tracking): remove old posthog/models/error_tracking directory

Models have been fully migrated to products/error_tracking/backend/models.
The old directory structure is no longer needed.

webjunkie committed:
fix(error_tracking): update stale import in error_tracking_query_runner

Import ErrorTrackingIssueSerializer from products.error_tracking.backend.api instead of posthog.api

webjunkie committed:
merge: adopt fresh migration validation branch

Merges chore/models-migrations-error-tracking-fresh into PR branch to replace all content with validated migration output.

Fresh branch was created from latest master (5d611bd6cc) and ran entire migration workflow from scratch, confirming all manual fixes were needed and migration is correct.

All files replaced to match fresh validation output.

github-actions[bot] committed:
Update query snapshots

webjunkie committed:
fix(tests): move source.js.map fixture to new error_tracking location

Source map fixture needs to be in products/error_tracking/backend/api/test/fixtures/ after API migration

webjunkie committed:
chore: update mypy baseline after error_tracking migration

Update baseline to reflect moved files:
- posthog/api/error_tracking.py → products/error_tracking/backend/api/error_tracking.py
- Add migration file type annotations needed

webjunkie committed:
fix(hogai): properly flush test data in async context

Fix test failures in hogai onboarding tests caused by incorrectly
calling flush_persons_and_events() from async context.

The test setup creates persons and events for testing but needs to flush
them to the database. Since _set_up_onboarding_tests is async, wrap the
synchronous flush_persons_and_events() call with sync_to_async.

This line was accidentally introduced in an earlier commit that was
incompletely reverted.

webjunkie committed:
chore: merge master into error_tracking migration branch

Resolve conflicts in api imports and migration numbering. Add EvaluationRunViewSet
from master to llm_analytics imports while keeping error_tracking import.

webjunkie committed:
chore: renumber migration to 0879

Rename 0878_migrate_error_tracking_models to 0879 to avoid conflict with
master's 0878_alter_externaldatasource_source_type. Update dependencies
in both posthog and products migrations.

webjunkie committed:
chore: update mypy baseline after merge

webjunkie committed:
fix: restore test fixtures to original formatting

Lint-staged reformatted test fixture files during merge commit, but these
should remain in their original format as they are test data. Restore from
master state.

Commit:
Merge branch 'master' into chore/models-migrations-error-tracking

Commit:
fix

daibhin reviewed (approved):
Had to ship one fix because the import in the query runner was broken. Tested locally though and everything seems to work. Confirmed that issue data wasn't lost 🚢🚢🚢
